Update Crate Dependencies + posslibly fixed deserialization issue with items

ju6ge opened the pull request:
Hi,

i adjusted the code to be compatible with the newest version of reqwest and serde ;)

Kind regrads,
ju6ge

ju6ge committed:
update dependencies

ju6ge committed:
remove unnecessary mutables

ju6ge committed:
simplified serde include

Commit:
possibly fixed issue #2 💪

ju6ge commented:
After some more debugging if have found some more inconsistencys with the behavior of the zotero api and fixed them. 
Possibly fixes #2

ju6ge committed:
added derive crate and moved to workspace repo structure

ju6ge committed:
handle api pageination

ju6ge committed:
more useful functions

ju6ge committed:
added more convenience functions to Item type \o/

ju6ge committed:
fuck zoteros inconsistent date strings

ju6ge commented:
Hey,

it is a bit sad when projects die because the people involved do not have time anymore. Nobody should be on the hook for maintaining something if they have other things to do in life. 

So I like to ask if you would be willing to let me take over maintainership of this repositoy/crate? I have active use for it and would like to make sure that other people can too.

Thanks for creating it, it has saved me a lot of time.

Kind regards,
ju6ge

joshkarlin commented:
yo @Eonm, are you willing to agree to the above?  I would also like to use this crate and if @ju6ge is willing to maintain it then seems like everyone will be happy

ju6ge commented:
@joshkarlin I just wrote an email directly to the maintainer. Maybe email notifications where turned off. But the website in the profile is also offline. So maybe something happend? I hope not!

Is there a process in Github on how to deal with stuff like this? Or is the answer just hard fork it and create an new crate?

joshkarlin commented:
yeah i hope all is well.

i'm not sure any github processes are available...  from [this discussion](https://github.com/orgs/community/discussions/13470), it looks like it's still an open issue.  i suspect hard fork would be the easiest and quickest

Eonm commented:
Hello @ju6ge thank you for your work, I reviewed your pull request

My apologies,

I wasn't notified by GitHub and haven't checked my personal GitHub account in a while. I will review your pull request as soon as possible.

I didn't expect my library to be used by others. I need help to maintain this library, I will add you as maintainer of the project.

Best regards

Eonm commented on zotero/Cargo.toml:
```diff
@@ -0,0 +1,22 @@
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+reqwest = { version = "0.11.11", features = ["blocking", "json"] }
```
Perhaps we should use the asynchronous version of reqwest in combination with tokio. Or do it in another pull request

ju6ge replied:
I was thinking the same … but I would argue that a transition to async code should be a seperate pull request!

Eonm commented on zotero/src/data_structure/item/item_data/bill.rs:
```diff
@@ -1,11 +1,13 @@
 /// A proposed piece of legislation.
-#[derive(Default, Deserialize, Serialize, Clone, Debug, Builder)]
+#[derive(Default, Deserialize, Serialize, Clone, Debug, Builder, ItemCommon)]
 // #[serde(rename_all(deserialize = "camelCase", serialize = "camelCase"))]
```
We should remove this line

```suggestion
```

Eonm commented on zotero/src/data_structure/mod.rs:
```diff
@@ -1,11 +1,13 @@
 
 /// Transform a data structure to a json object
-pub trait ToJson: serde::Serialize {
+pub trait ToJson: Serialize {
     fn to_json(&self) -> String {
         serde_json::to_string(&self).unwrap()
```
My initial implementation is a bit dirty.

We should remove unwrap from code base.

May be `use std::convert::TryInto;` is more idiomatic than a custom trait ?

ju6ge replied:
I am not sure we really need to explicitly provide an ToJson implementation. Serde handles this well, and a library user can do the conversion without any issue and handle it.

Eonm commented on zotero/src/data_structure/shared_fields.rs:
```diff
@@ -24,6 +23,32 @@ pub struct Link {
+                return true;
+            }
+        }
+        return false;
```
Personally, I prefer a more functional style

```suggestion
      self.tags().iter().any(|t| t.tag == tag)
```

Eonm commented on zotero/src/data_structure/item/mod.rs:
```diff
@@ -192,6 +202,289 @@ pub struct Item {
+                return true;
+            }
+        }
+        return false;
```
```suggestion
self.tags().iter().any(|t| t.tag == tag)
```

Eonm commented on zotero/src/data_structure/item/mod.rs:
```diff
@@ -204,17 +497,87 @@ pub struct Creator {
+        if self.first_name.len() == 0 {
+            return format!("{}", self.last_name)
+        }
+        format!("{}. {}", self.first_name.chars().next().unwrap(), self.last_name)
```
```suggestion
match self.first_name.chars().next() {
    Some(first_initial) => format !("{}. {}", first_initial, self.last_name),
    None => format !("{}", self.last_name),
}
```
This code doesn't work for some names: `Jean-Philippe` should be transformed into `J.-P.` or `Hanna Mary` into `H. M.`

Perhaps we should declare the surname as optional and set the value to None when deserializing an empty string?

ju6ge replied:
i changed the code to the suggest logic, a more though update would be needed to enable smarter handling of names, we should probably write some test cases to to ensure that it is handled correctly

Eonm commented on zotero/src/data_structure/item/mod.rs:
```diff
@@ -204,17 +497,87 @@ pub struct Creator {
+}
+
+#[derive(Clone, Debug, Serialize, Deserialize, PartialEq)]
+pub enum SizeOrBool {
```
```suggestion
#[serde(untagged)]
pub enum SizeOrBool {
```

With `#[serde(untagged)]` serde will try to deserialize against each variant.

Eonm commented on zotero/src/data_structure/item/mod.rs:
```diff
@@ -204,17 +497,87 @@ pub struct Creator {
+}
+
+/// A custom deserializer that deserialize parent_collection value either in bool or in Size.
+fn deserialize_meta_num_children<'de, D>(deserializer: D) -> Result<Option<SizeOrBool>, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let s: Value = serde::Deserialize::deserialize(deserializer)?;
+    if s.is_boolean() {
+        match serde_json::from_value::<bool>(s) {
+            Err(_) => Ok(None),
+            Ok(v) => Ok(Some(SizeOrBool::Bool(v)))
+        }
+    } else if s.is_number() {
+        match serde_json::from_value::<usize>(s) {
+            Err(_) => Ok(None),
+            Ok(v) => Ok(Some(SizeOrBool::Size(v)))
+        }
+    } else if s.is_null() {
+        Ok(None)
+    } else {
+        panic!("invalid type")
+    }
+}
+
+
```
```suggestion
```

This code might not be necessary if untagged works (https://serde.rs/enum-representations.html#untagged)

ju6ge replied:
You are right this makes it simpler, no custom deserialize function required.

Eonm commented on zotero/src/data_structure/item/mod.rs:
```diff
@@ -192,6 +202,289 @@ pub struct Item {
+}
+
+fn convert_zotero_date_str(date_str: &str) -> DateTime<Local> {
+    let date_regex = Regex::new(r"(\d{4})-(\d{2})-(\d{2})").unwrap();
```
We should move this regex inside a lazy_static to avoid building the regex each time the function is called.

https://stackoverflow.com/questions/35169259/how-to-make-a-compiled-regexp-a-global-variable

Eonm commented on zotero/src/data_structure/item/mod.rs:
```diff
@@ -192,6 +202,289 @@ pub struct Item {
+    let date_regex = Regex::new(r"(\d{4})-(\d{2})-(\d{2})").unwrap();
+    let date_captures = date_regex.captures(date_str);
+
+    let formatter_regex = Regex::new(r"(\d{2})/(\d{4})").unwrap();
```
We should move this regex inside a lazy_static to avoid building the regex each time the function is called.

https://stackoverflow.com/questions/35169259/how-to-make-a-compiled-regexp-a-global-variable

ju6ge committed:
Update zotero/src/data_structure/shared_fields.rs

Co-authored-by: Eonm <eon.mathis@gmail.com>

ju6ge committed:
Update zotero/src/data_structure/item/mod.rs

Co-authored-by: Eonm <eon.mathis@gmail.com>

ju6ge committed:
Update zotero/src/data_structure/item/item_data/bill.rs

Co-authored-by: Eonm <eon.mathis@gmail.com>

ju6ge commented:
@Eonm great to hear from you. I will gladly help with the maintenance of this project ;) 

For now I did some quick review of the updates you suggested. For the rest I need some more time ;)

Eonm commented:
@ju6ge Thanks for the changes.

Just one last thing, one of the unit tests doesn't pass. Once fixed, I'll merge the pull request

ju6ge committed:
updated test to pass

ju6ge committed:
moved regex to static variable

ju6ge committed:
use serde(untagged) instead of custom deserializer

ju6ge committed:
fixed compiler warnings

ju6ge committed:
chore: cargo fmt

ju6ge committed:
chore: cargo clippy

ju6ge committed:
apply change to short name function, does not change the logic

ju6ge commented:
So I updated the PR and cleaned up the code base with `cargo fmt` and `cargo clippy`. All test should be passing now. The next thing I will be working on is enabling async support. But that should be it's own pull request.
